File: prime.cpp
#include <cstdio>
#include <cmath>
#define MAX 220


using namespace std;

bool sieve[MAX];
bool val[MAX];//false = 0 true = 1 NUMBER OF TIMES CONTAINED MOD 2
bool test[MAX];
int in;

void factor(int origin){
    int input = origin;
    int start = 2;
    while(start<=input){
        while(input%start==0){
            input/=start;
            val[start] = !val[start];
        }
        do {start++;}
        while(sieve[start]);
    }
    if(input!=1) printf("%d ERROR\n",origin); 
}
void fill(int origin){
    int input = origin;
    int start = 2;
    while(start<=input){
        while(input%start==0){
            input/=start;
            test[start] = !test[start];
        }
        do {start++;}
        while(sieve[start]);
    }
    if(input!=1) printf("%d ERROR\n",origin); 

}
void print(){
    for(int x = 2;x<MAX;x++){
        if(!sieve[x]){
            printf("PRIME:%d NUMBER:%d\n",x,val[x]);
            
        }
    }
}
void compare(int value){
    for(int x = 0;x<MAX;x++){
        if(val[x]!=test[x]) return;
    }
    printf("%d\n",value);
}

int main(){
    
    int stop = sqrt(MAX);
    sieve[0] = true;
    sieve[1] = true;
    for(int x = 2;x<stop;x++){
        if(sieve[x]==false){//we are prime
            for(int a = x*x;a<MAX;a+=x){
                sieve[a] = true;//we are composite
            }
        }
    }
    for(int x = 2;x<=200;x+=2){
        factor(x);
    } 
    print();
    for(int x = 2;x<=96;x++) fill(x);
    for(int x = 97;x<211;x++){
        fill(x);
        compare(x);
    }
    /*while(true){
        
        scanf("%d",&in);
        if(in==-1) break;
        factor(in);
        print(); 
    }*/
    


    return 0;
}

File: Prime.java
package Algorithms;
import java.util.*;

/*Intersting facts about primes INCOMPLETE

All primes greater than three can be written as 1 or -1 mod 6.
Sieve of Eratosthenes is standard

Using Fermat's little theorem and a strong primality test is the basis of RSA

*/

public class Prime {
	static ArrayList<Integer> primes;
	static boolean[] sieve;//sieve of composites
	
	public static void main(String[] args){
		primes = new ArrayList<Integer>();
		primes.add(2);
		for(int x = 3;x<Math.sqrt(600851475143L);x++){
			if(isPrime(x)) primes.add(x);
		}
		
		System.out.println(factor(600851475143L));
		
	}
	public static void sieve(int max){//SIEVE DOES NOT WORK!!!!!
		int limit = (int)Math.sqrt(max);
		int cur = 3;
		int marker = -1;
		while(cur<=limit){
			marker = (3*cur-3)/2;
			while(marker<=max){
				sieve[marker] = true;
				marker+=cur;
			}
			for(int x = (cur-1)/2;x<max;x++){
				if(!sieve[x]) cur = 2*x+3;
			}
		}
	}
	
	public static boolean isPrime(long test){//Standard Prime Tester
		long max = (long)Math.sqrt(test);
		int counter = 0;
		long prime = primes.get(0);
		while(prime<=max){
			if(test%prime==0) return false;
			counter++;
			prime = primes.get(counter);
		}
		return true;
	}
	public static long factor(long test){//returns the largest prime factor
		
		long max = (long)Math.sqrt(test);
		int counter = 0;
		long prime = primes.get(0);
		while(prime<=max){
			while(test%prime==0) test/=prime;
			counter++;
			prime = primes.get(counter);
			max = (long)Math.sqrt(test);
		}
		return test;
	}
	
	
	
}
